identity/src/main/java/com/android/identity/internal/Util.kt

```diff
@@ -0,0 +1,2209 @@
+        val protectedHeaders = CborBuilder()
+        val protectedHeadersMap: MapBuilder<CborBuilder> = protectedHeaders.addMap()
+        protectedHeadersMap.put(COSE_LABEL_ALG, COSE_ALG_HMAC_256_256)
+        val protectedHeadersBytes = cborEncode(protectedHeaders.build().get(0))
+        val toBeMACed = coseBuildToBeMACed(protectedHeadersBytes, data!!, detachedContent!!)
+        val mac: ByteArray
+        mac = try {
+            val m = Mac.getInstance("HmacSHA256")
+            m.init(key)
+            m.update(toBeMACed)
+            m.doFinal()
+        } catch (e: NoSuchAlgorithmException) {
+            throw IllegalStateException("Unexpected error", e)
+        } catch (e: InvalidKeyException) {
+            throw IllegalStateException("Unexpected error", e)
+        }
+        val builder = CborBuilder()
+        val array: ArrayBuilder<CborBuilder> = builder.addArray()
+        array.add(protectedHeadersBytes)
+        /* MapBuilder<ArrayBuilder<CborBuilder>> unprotectedHeaders = */array.addMap()
+        if (data.size == 0) {
+            array.add(SimpleValue(SimpleValueType.NULL))
+        } else {
+            array.add(data)
+        }
+        array.add(mac)
+        return builder.build().get(0)
+    }
+
+    @JvmStatic
+    fun coseMac0GetTag(coseMac0: DataItem): ByteArray {
+        val items = castTo(
+            co.nstant.`in`.cbor.model.Array::class.java, coseMac0
+        ).dataItems
+        require(items.size >= 4) { "coseMac0 have less than 4 elements" }
+        val tagItem = items[3]
+        return castTo(
+            ByteString::class.java, tagItem
+        ).bytes
+    }
+
+    /**
+     * Brute-force but good enough since users will only pass relatively small amounts of data.
+     */
+    @JvmStatic
+    fun hasSubByteArray(haystack: ByteArray, needle: ByteArray): Boolean {
+        var n = 0
+        while (needle.size + n <= haystack.size) {
+            var found = true
+            for (m in needle.indices) {
+                if (needle[m] != haystack[n + m]) {
+                    found = false
+                    break
+                }
+            }
+            if (found) {
+                return true
+            }
+            n++
+        }
+        return false
+    }
+
+    @JvmStatic
+    fun stripLeadingZeroes(value: ByteArray): ByteArray {
+        var n = 0
+        while (n < value.size && value[n].toInt() == 0) {
+            n++
+        }
+        val newLen = value.size - n
+        val ret = ByteArray(newLen)
+        var m = 0
+        while (n < value.size) {
+            ret[m++] = value[n++]
+        }
+        return ret
+    }
+
+    /**
+     * Returns #6.24(bstr) of the given already encoded CBOR
+     */
+    @JvmStatic
+    fun cborBuildTaggedByteString(encodedCbor: ByteArray): DataItem {
+        val item: DataItem = ByteString(encodedCbor)
+        item.setTag(CBOR_SEMANTIC_TAG_ENCODED_CBOR)
+        return item
+    }
```

Comment:
using `apply{ ..}` allows us to access the object and perform additions to it then return the object (with all the changes)

```
fun cborBuildTaggedByteString(encodedCbor: ByteArray): DataItem =
        ByteString(encodedCbor).apply { 
            setTag(CBOR_SEMANTIC_TAG_ENCODED_CBOR)
        }
```